HW1/P5/P5_bfs.py

```diff
@@ -0,0 +1,81 @@
+
+import pyspark
+from pyspark import SparkContext
+sc = SparkContext()
+
+
+# read in data
+#links = sc.textFile('s3://Harvard-CS205/wikipedia/links-simple-sorted.txt', 32)
+#page = sc.textFile('s3://Harvard-CS205/wikipedia/titles-sorted.txt', 32)
+
+links = sc.textFile('links-simple-sorted.txt')
+page = sc.textFile('titles-sorted.txt')
+
+# construct rdd of all neighbors
+rdd1 = links.map(lambda x: x.split(': '))
+def get_dest(s):
+    return [int(x) for x in s.split(' ')]
+rdd2 = rdd1.map(lambda x: (int(x[0]), get_dest(x[1])))
+
+# partition the graph rdd
+rdd_link = rdd2.partitionBy(256).cache()
+
+# set index to page names and get Kevin Bacon, Harvard index
+rdd3 = page.zipWithIndex()
+rdd_page = rdd3.map(lambda x: (x[1]+1, x[0])).cache()
+
+# get index of start and end nodes
+start = rdd_page.filter(lambda x: x[1] == 'Kevin_Bacon').take(1)[0][0]
+end = rdd_page.filter(lambda x: x[1] == 'Harvard_University').take(1)[0][0]
+
+
+def bfs(start, end, rdd_link, rdd_page):
+    find_path = sc.accumulator(0)
+    find_path.add(1)
+    # keep track of path
+    path = []
+    # keep track of current layer of nodes
+    current = [start]
+    l = [start]
+    index = 1
+
+
+    while find_path.value != 0:
+        find_path.value = 0
+        this_layer = rdd_link.filter(lambda x: x[0] in current).flatMap(lambda x: x[1]).distinct().map(lambda x: (index,x))
+        path = path + this_layer.collect()
+        current = list(set(this_layer.values().collect()) - set(l))
```

Comment:
Most of the actions here are done in python and not in spark (using python's sets and lists).